### PR TITLE
Mg replication refactor part 6 again

### PR DESCRIPTION
**Pull request by andrejtonev**
[master < Epic] PR
- [ ] Check, and update documentation if necessary
- [ ] Write E2E tests
- [ ] Compare the [benchmarking results](https://bench-graph.memgraph.com/) between the master branch and the Epic branch
- [ ] Provide the full content or a guide for the final git message

[master < Task] PR
- [ ] Check, and update documentation if necessary
- [ ] Provide the full content or a guide for the final git message

To keep docs changelog up to date, one more thing to do:
- [ ] Write a release note here, including added/changed clauses
- [ ] Tag someone from docs team in the comments

### CODE REVIEW

File: src/storage/v2/replication/recovery.hpp [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+#pragma once
+
+#include <filesystem>
+#include <variant>
+#include <vector>
+
+using RecoverySnapshot = std::filesystem::path;
```
Comment (Ignition) [inline src/storage/v2/replication/recovery.hpp:18]: put inside a namespace

File: src/storage/v2/replication/replication_client.hpp [hunk 2]
```diff
@@ -61,85 +66,80 @@ class ReplicaStream {
...
+  ~ReplicationStorageClient() = default;
 
-  auto Mode() const -> memgraph::replication::ReplicationMode { return mode_; }
-  auto Name() const -> std::string const & { return name_; }
-  auto Endpoint() const -> io::network::Endpoint const & { return rpc_client_.Endpoint(); }
-  auto State() const -> replication::ReplicaState { return replica_state_.load(); }
-  auto GetTimestampInfo() -> TimestampInfo;
+  auto Mode() const -> memgraph::replication::ReplicationMode { return client_.mode_; }
+  auto Name() const -> std::string const & { return client_.name_; }
+  auto Endpoint() const -> io::network::Endpoint const & { return client_.rpc_client_.Endpoint(); }
```
Comment (Ignition) [inline src/storage/v2/replication/replication_client.hpp:96]: TODO: reduce/remove need to access these from `ReplicationStorageClient`

File: src/storage/v2/replication/replication_client.hpp [hunk 3]
```diff
@@ -61,85 +66,80 @@ class ReplicaStream {
...
+ private:
+  void RecoverReplica(uint64_t replica_commit, memgraph::storage::Storage *storage);
+
+  void CheckReplicaState(Storage *storage);
+  void LogRpcFailure();
+  void TryCheckReplicaStateSync(Storage *storage);
+  void FrequentCheck(Storage *storage);
+
+  ::memgraph::replication::ReplicationClient &client_;
+  std::optional<ReplicaStream>
```
Comment (Ignition) [inline src/storage/v2/replication/replication_client.hpp:139]: TODO: make granted streams local variables, rather than stored here

Comment (andrejtonev) [reply src/storage/v2/replication/replication_client.hpp:139]: Not sure this is actually possible/advisable. The stream is a crucial part of the state machine.

File: src/dbms/replication_handler.cpp [hunk 4]
```diff
@@ -166,15 +181,15 @@ auto ReplicationHandler::UnregisterReplica(std::string_view name) -> UnregisterR
...
     }
-    auto const n_unregistered =
-        std::erase_if(mainData.registered_replicas_,
-                      [&](ReplicationClientConfig const &registered_config) { return registered_config.name == name; });
-
-    dbms_handler_.ForEach([&](Database *db) {
-      db->storage()->repl_storage_state_.replication_clients_.WithLock(
-          [&](auto &clients) { std::erase_if(clients, [&](const auto &client) { return client->Name() == name; }); });
+    // Remove database specific clients
+    dbms_handler_.ForEach([name](Database *db) {
```
[the remark this answers is not in the document]
Comment (andrejtonev) [reply src/dbms/replication_handler.cpp:185]: Do you think this falls under the already known issue of ReplicationState synchronization?

[the remark this answers is not in the document]
Comment (andrejtonev) [reply src/dbms/replication_handler.cpp:190]: Next PR fix the ReplicationState sync

File: src/storage/v2/replication/replication_client.hpp (hunk 3 above)
Comment (andrejtonev) [reply src/storage/v2/replication/replication_client.hpp:142]: Can be done, will be done later on.

File: src/storage/v2/durability/durability.cpp [hunk 5]
```diff
@@ -106,20 +107,25 @@ std::optional<std::vector<WalDurabilityInfo>> GetWalFiles(const std::filesystem:
 
   std::vector<WalDurabilityInfo> wal_files;
   std::error_code error_code;
+  // There could be multiple "current" WAL files, the tah just means that the previous session didn't finalize.
```
Comment (gvolfing) [inline src/storage/v2/durability/durability.cpp:110]: what's `tah`?

**Review by gvolfing — COMMENTED** [review]
Not much to comment, good job.
I don't want to reiterate it this too much but we will really need to update the internal documentation about this to make it digestible.

**Comment by andrejtonev** [comment]
> Not much to comment, good job. I don't want to reiterate it this too much but we will really need to update the internal documentation about this to make it digestible.

We have been adding documentation to notion. Looks for "Replication (v2.13.0)". I think we cover most of the things, but could be missing something just from being to deep in it.

File: src/storage/v2/durability/durability.cpp (hunk 5 above)
Comment (andrejtonev) [reply src/storage/v2/durability/durability.cpp:110]: "tag" :smile:

File: src/dbms/dbms_handler.cpp [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+// licenses/APL.txt.
+
+#include "dbms/dbms_handler.hpp"
+
+namespace memgraph::dbms {
+#ifdef MG_ENTERPRISE
+DbmsHandler::DbmsHandler(
+    storage::Config config,
+    memgraph::utils::Synchronized<memgraph::auth::Auth, memgraph::utils::WritePrioritizedRWLock> *auth,
+    bool recovery_on_startup, bool delete_on_drop)
```
[the remark this answers is not in the document]
Comment (andrejtonev) [reply src/dbms/dbms_handler.cpp:19]: Added a comment

**Review by gvolfing — APPROVED** [review]

**Comment by andrejtonev** [comment]
@vpavicic 

In part 5 and part 6 we added several hidden multi-tenant/replication features.
The features are behind a compile-time feature flag and were meant to not be broadcast to users until completely finished.

Now we want to have the feature in the docs, even if it is hidden behind the feature flag.

If compiled with the `MG_EXPERIMENTAL_REPLICATION_MULTITENANCY = ON` we can replicate multi-tenant data.
We are still not able to replicate CREATE/DROP of the database themselves. 

If the user wants to have multi-tenant replication (and has Memgraph which is compiled correctly), they need to:
1. Setup the databases on all instances while each one is still in MAIN mode.
2. Switch replicas to REPLICA mode and connect to MAIN, as one would normally do.
3. At this point creating a new database or connecting a replica that is missing a database will fail. Dropping a database is disabled on replicas, but is possible on MAIN, although dangerous. We advise to not drop databases once everything is connected.

Not sure if this should go under replication documentation or a changelog. In any case, this is all highly experimental.

**Comment by Ignition** [comment]
related to #1359

**Pull request merged**